import { Test } from '@nestjs/testing';
import { UserRepository } from './user.repository';
import { ConflictException, InternalServerErrorException } from '@nestjs/common';
import { User } from './user.entity';
import * as faker from 'faker';
import * as bcrypt from 'bcryptjs';

const mockCredentialsDto = {
  username: faker.internet.userName,
  password: faker.internet.password(8),
  hashPassword: faker.internet.password(20)
};

describe('UserRepository', () => {
  let userRepository;

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [
        UserRepository
      ]
    }).compile();

    userRepository = await module.get<UserRepository>(UserRepository);
  });

  describe('signUp', () => {
    let save;

    beforeEach(() => {
      save = jest.fn();
      userRepository.create = jest.fn().mockReturnValue({ save });
    })

    test('successfully signs up the user', () => {
      save.mockResolvedValue(undefined);
      expect(userRepository.signUp(mockCredentialsDto)).resolves.not.toThrow();
    });

    test('throws a conflict exception as username already exists', () => {
      save.mockRejectedValue({ code: '23505' });
      expect(userRepository.signUp(mockCredentialsDto)).rejects.toThrow(ConflictException);
    });

    test('throws a conflict exception as username already exists', () => {
      save.mockRejectedValue({ code: faker.random.number });
      expect(userRepository.signUp(mockCredentialsDto)).rejects.toThrow(InternalServerErrorException);
    });
  });

  describe('validateUserPassword', () => {
    let user;

    beforeEach(() => {
      userRepository.findOne = jest.fn();
      user = new User();
      user.username = faker.internet.userName;
      user.validatePassword = jest.fn();
    });

    test('returns the username as validation is successful', async () => {
      userRepository.findOne.mockResolvedValue(user);
      user.validatePassword.mockResolvedValue(true);

      const result = await userRepository.validateUserPassword(mockCredentialsDto);
      expect(result).toEqual(user.username);
    });

    test('returns null as user cannot be found', async () => {
      userRepository.findOne.mockResolvedValue(null);
      
      const result = await userRepository.validateUserPassword(mockCredentialsDto);
      expect(user.validatePassword).not.toHaveBeenCalled();
      expect(result).toBeNull();
    });

    test('returns null as password is invalid', async () => {
      userRepository.findOne.mockResolvedValue(user);
      user.validatePassword.mockResolvedValue(false);

      const result = await userRepository.validateUserPassword(mockCredentialsDto);
      expect(user.validatePassword).toHaveBeenCalled();
      expect(result).toBeNull();
    });
  });

  describe('hashPassword', () => {
    it('calls bcrypt.hash to generate a hash', async () => {
      bcrypt.hash = jest.fn().mockResolvedValue(mockCredentialsDto.hashPassword);
      expect(bcrypt.hash).not.toHaveBeenCalled();

      const salt = await bcrypt.genSalt();
      const result = await userRepository.hashPassword(mockCredentialsDto.password, salt);
      expect(bcrypt.hash).toHaveBeenCalledWith(mockCredentialsDto.password, salt);
      expect(result).toEqual(mockCredentialsDto.hashPassword);
    });
  });
  
});
